codecs/text/src/main/java/com/datastax/oss/dsbulk/codecs/text/string/StringToVectorCodec.java

```diff
@@ -17,18 +17,42 @@
 
 import com.datastax.oss.driver.api.core.data.CqlVector;
 import com.datastax.oss.driver.internal.core.type.codec.VectorCodec;
+import com.datastax.oss.driver.shaded.guava.common.base.Splitter;
+import com.datastax.oss.driver.shaded.guava.common.collect.Streams;
+import com.datastax.oss.dsbulk.codecs.api.ConvertingCodec;
+import java.util.ArrayList;
 import java.util.List;
+import java.util.stream.Collectors;
 
 public class StringToVectorCodec<SubtypeT extends Number>
     extends StringConvertingCodec<CqlVector<SubtypeT>> {
 
-  public StringToVectorCodec(VectorCodec<SubtypeT> targetCodec, List<String> nullStrings) {
+  private final ConvertingCodec<String, SubtypeT> stringCodec;
+
+  public StringToVectorCodec(
+      VectorCodec<SubtypeT> targetCodec,
+      ConvertingCodec<String, SubtypeT> stringCodec,
+      List<String> nullStrings) {
     super(targetCodec, nullStrings);
+    this.stringCodec = stringCodec;
   }
 
   @Override
   public CqlVector<SubtypeT> externalToInternal(String s) {
-    return this.internalCodec.parse(s);
+
+    // Logic below adapted from VectorCodec.parse() and CqlVector.from() but here we use the dsbulk
+    // codecs for the subtype in order to enforce any additional behaviours
+    //
+    // Logically this probably makes more sense anyway.  It's the responsibilty of dsbulk to define
+    // what sorts of formats it wants to support for vectors.  It can certainly re-use a
+    // representation
+    // known to work with vectors but it certainly isn't obligated to do so.
+    if (s == null || s.isEmpty() || s.equalsIgnoreCase("NULL")) return null;
+    ArrayList<SubtypeT> vals =
```

Comment:
This doesn't look 100% right to me.

If you are going down the path of having a subcodec – which I think is the right thing to do btw – then I would suggest to go even further: here, you are using a string codec to parse N elements of the vector. Because of that, you need to handle the collection manually, by stripping the first and last character of the input, which are assumed to be the collection delimiters.

But instead, I think you should use a _collection codec_ that would parse the _entire_ collection for you. That collection codec would in turn use another inner codec to parse the elements.

Reply:
If I understand correctly you're thinking of something more like this then?

```java
public class StringToVectorCodec<SubtypeT extends Number>
    extends StringToCollectionCodec<SubtypeT, CqlVector<SubtypeT>> {
...
}
```

If so then we have an issue with the underlying Java type: CqlVector [implements Iterable but doesn't implement Collection](https://github.com/apache/cassandra-java-driver/blob/4.18.1/core/src/main/java/com/datastax/oss/driver/api/core/data/CqlVector.java#L50-L55), so it can't satisfy the type bounds in question.  The rationale here was that CqlVector is closer to an array than a Collection type so implementing Collection would be somewhat confusing.  I readily agree there's room for debate here but it is the state of the world as of 4.18.1.

Does that fact change your analysis?  It seems like there are at least three paths forward here:

1. Make CqlVector implement Collection.  Would require a new Java driver release + a dsbulk update to use the new version.
2. Adapt the JSON parsing logic in StringToCollectionCodec to be used here instead of the manual process.  Provides similar parsing behaviour to what dsbulk currently exhibits for collections without the need for a driver upgrade at the cost of some code duplication.
3. Leave it as-is, possibly with a ticket for future work

I'm not opposed to any of these options, although (1) will certainly take more time than the other two.

Thoughts?

Reply:
@absurdfarce here is what I had in mind:

https://github.com/datastax/dsbulk/compare/issue484...adutra:dsbulk:issue484

It's loosely inspired by how `StringToMapCodec` is designed.

Also, it adheres to the general principle in dsbulk that mandates that all complex codecs rely in a standard _json_ representation of the data, and not their CQL representation. So here `StringToVectorCodec` relies on a json codec that outputs a json array of the vector dimensions.

Let me know if that looks better.

Reply:
**Ohhhhh**, I see what you mean now.  Yeah, this makes _a ton_ more sense than my original suggestion!  I'm gonna make a PR out of that diff and merge it in here; there are several improvements in there I'd like to get into this work.

Many thanks, my friend; that diff absolutely crystalized exactly what you were talking about!